### PR TITLE
Add shmem_signal_{add,set}

### LINKED ISSUE #382: Add Signal API
## Issue

Signaling is used for point-to-point synchronization. With OpenSHMEM 1.5 we removed support for `shmem_p` as a signal operation.

Current replacements for `shmem_p` are `shmem_set` (or another AMO) or zero-length `shmem_put_signal`. Atomic set must be atomic with respect to all other AMOs; signals can provide better performance (only require single-copy atomicity).

Put-with-signal can be used for only transmitting a signal update, but for this usage model the API is not very clean and there are additional overheads (arguments, branches, etc.) compared with a simple signal operation.

## Proposed Solution

Add a signal operation:

`void shmem_signal(uint64_t *sig_addr, uint64_t signal, int sig_op, int pe);`

### PR DESCRIPTION
**Pull request by nspark**
- Adds `shmem_signal_{add,set}`
- Revises the signaling operations introduction and atomicity sections accordingly

Closes #382

### CODE REVIEW

**Comment by nspark** [comment]
Yes. AMOs are only guaranteed atomicity of updates w.r.t. other AMOs of the same data type (but for all operation types; e.g., set, add, xor). Signal operations are only guaranteed atomicity of updates w.r.t. other signal operations of the same operation type (i.e., add or set).

I think the intent was that, for performance reasons, `shmem_signal_set` and `shmem_put_signal(SET)` can provide (for example) single-copy atomic writes without needing the full strength of atomicity required by AMOs.

**Comment by shamisp** [comment]
I think this adds quite a bit API bloat, while the semantical difference is very nuanced. I would suggest to add a parameter to shmalloc that lets you define desired AMO semantics, so you can reuse all existing AMO operations.

File: content/shmem_signal_add.tex [hunk 1]
```diff
@@ -0,0 +1,47 @@
+\apisummary{
+  Adds to a signal value of a remote data object.
+}
+
+\begin{apidefinition}
+
+\begin{C11synopsis}
+void @\FuncDecl{shmem\_signal\_add}@(shmem_ctx_t ctx, const uint64_t *sig_addr, uint64_t signal, int pe);
```
Comment (nspark) [inline content/shmem_signal_add.tex:8]:
```suggestion
void @\FuncDecl{shmem\_signal\_add}@(shmem_ctx_t ctx, uint64_t *sig_addr, uint64_t signal, int pe);
```

File: content/shmem_signal_set.tex [hunk 2]
```diff
@@ -0,0 +1,47 @@
+\apisummary{
+  Sets the signal value of a remote data object.
+}
+
+\begin{apidefinition}
+
+\begin{C11synopsis}
+void @\FuncDecl{shmem\_signal\_set}@(shmem_ctx_t ctx, const uint64_t *sig_addr, uint64_t signal, int pe);
```
Comment (nspark) [inline content/shmem_signal_set.tex:8]:
```suggestion
void @\FuncDecl{shmem\_signal\_set}@(shmem_ctx_t ctx, uint64_t *sig_addr, uint64_t signal, int pe);
```

File: content/shmem_signal_set.tex [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+\end{C11synopsis}
+
+\begin{Csynopsis}
+void @\FuncDecl{shmem\_signal\_set}@(const uint64_t *sig_addr, uint64_t signal, int pe);
+void @\FuncDecl{shmem\_ctx\_signal\_set}@(shmem_ctx_t ctx, const uint64_t *sig_addr, uint64_t signal, int pe);
```
Comment (nspark) [inline content/shmem_signal_set.tex:13]:
```suggestion
void @\FuncDecl{shmem\_signal\_set}@(uint64_t *sig_addr, uint64_t signal, int pe);
void @\FuncDecl{shmem\_ctx\_signal\_set}@(shmem_ctx_t ctx, uint64_t *sig_addr, uint64_t signal, int pe);
```

**Review by nspark — COMMENTED** [review]
Remove erroneous `const` type qualifiers

**Comment by manjugv** [comment]
May spec meeting: Please add change log entry

**Comment by jdinan** [comment]
Special ballot: [1c254ce](https://github.com/openshmem-org/specification/pull/489/commits/1c254cea3d29ac3fec357eb2b9fbddd896152994)

**Pull request merged**